Testing and updating the Go Code Quick Start (#1956)
timescale/docs: timescaledb/quick-start/golang.md

---
title: "Quick Start: Go and TimescaleDB"
excerpt: Get started with TimescaleDB for a Go application
keywords: [Golang, Go]
---

import Intro from "versionContent/_partials/_quickstart-intro.mdx";

# Go quick start

<Intro />

This quick start guide walks you through:

*   [Connecting to TimescaleDB][connect]
*   [Creating a relational table][create-table]
*   [Creating a hypertable][create-hypertable]
*   [Inserting data][insert]
*   [Executing a query][query]

## Prerequisites

Before you start, make sure you have:

*   Installed TimescaleDB. For more information, see the
    [installation documentation][install].
*   Installed Go. For more information, see the [Go documentation][golang-install].
*   Installed the [PGX driver][pgx-driver-github] for Go.

<highlight type="cloud" header="Run all tutorials free" button="Try for free">
Your Timescale Cloud trial is completely free for you to use for the first
thirty days. This gives you enough time to complete all the tutorials and run
a few test projects of your own.
</highlight>

## Connect to TimescaleDB

> FIXME

Author: I was hoping you would help with some of this text 😉 

I'll take another walk through today.

> FIXME

Reviewer: Sure, I was not sure what exactly to look for there. I'll try to help too, I just didn't know :)

---
title: "Quick Start: Go and TimescaleDB"
excerpt: Get started with TimescaleDB for a Go application
keywords: [Golang, Go]
---

> import Intro from "versionContent/_partials/_quickstart-intro.mdx";

Author: It will be the same partial in all the quick starts, though, so it shouldn't matter.

> import Intro from "versionContent/_partials/_quickstart-intro.mdx";

Reviewer 2: Thinking about other pages that might want to have a partial named Intro though, like maybe some tutorials that might want to share it.

But if we use more specific names elsewhere, having one special case will not be a problem :)

# Go quick start

<Intro />

This quick start guide walks you through:

*   [Connecting to TimescaleDB][connect]
*   [Creating a relational table][create-table]
*   [Creating a hypertable][create-hypertable]
*   [Inserting data][insert]

> *   [Executing a query][query]

Reviewer 2: Except for 'creating a hypertable', these links don't work in preview

## Prerequisites

Before you start, make sure you have:

> *   Installed TimescaleDB. For more information, see the

Reviewer 2: Not sure what is wrong with this link, it thinks it's a relative link but I don't see anything wrong with the syntax... 🤔

Author: It had `/latest/` in it rather than `:currentVersion:`, but I also made it point to the Cloud install rather than the landing page.

    statement string you created:

    ```go
    package main

    import (
       "context"
       "fmt"
       "os"

       "github.com/jackc/pgx/v5"
       "github.com/jackc/pgx/v5/pgxpool"
    )

    func main() {

       ctx := context.Background()
       connStr := "yourConnectionStringHere"
       dbpool, err := pgxpool.New(ctx, connStr)
       if err != nil {
           fmt.Fprintf(os.Stderr, "Unable to connect to database: %v\n", err)
           os.Exit(1)
       }
       defer dbpool.Close()

       /********************************************/
       /* Create relational table                      */
       /********************************************/

       //Create relational table called sensors
       queryCreateTable := `CREATE TABLE sensors (id SERIAL PRIMARY KEY, type VARCHAR(50), location VARCHAR(50));`
       _, err = dbpool.Exec(ctx, queryCreateTable)
       if err != nil {
           fmt.Fprintf(os.Stderr, "Unable to create SENSORS table: %v\n", err)
           os.Exit(1)
       }
       fmt.Println("Successfully created relational table SENSORS")
    }
       ```

</Collapsible>

</procedure>

## Generate a hypertable

> Virtually all user interactions with TimescaleDB are with hypertables. Creating

Reviewer 2: Do we like this statement? I don't love it, I don't know how true it is, because you can have tons of relational tables for metadata as well

Author: Yeah, this is old text. It made it into the Python one too. I wonder if we can just delete it.

Reviewer 2: Totally passed me by on that way. Would be up for deleting it

tables and indexes, altering tables, inserting data, selecting data, and most
other tasks can and should all be executed on the hypertable.

<procedure>

<Collapsible heading="Generating a hypertable" headingLevel={3}>

1.  Create a variable for the `CREATE TABLE SQL` statement for your hypertable.
    Notice how the hypertable has the compulsory time column:

    ```go

>     FIXME

Author: @jonatas We still have a FIXME in here ... do you have something for this one?

    ```

1.  Formulate the `SELECT` statement to convert the table into a hypertable. You
    must specify the table name to convert to a hypertable, and its time column

>     name as the two. For more information, see the

Reviewer 2: ```suggestion
    name as the second argument. For more information, see the
```

    [`create_hypertable` docs][create-hypertable-docs]:

    ```go

>     queryCreateHypertable := `CREATE TABLE sensor_data (

Reviewer 2: I think this code block is mixing up the previous step and the current step. The last line belongs to this step, and the rest belongs to the previous step.

Author: @jonatas ?

Reviewer: Yes! that was a mess. I think I resolved the issue now.

Reviewer: Just to make it clear, it starts exposing how the command will be and later it wraps up the same code block into a larger context that is runnable.

I think this is right now.

       }

       // Check contents of results slice
       fmt.Println("Contents of RESULTS slice")
       for i := range results {
           var r result
           r = results[i]
           fmt.Printf("Time: %s | ID: %d | Temperature: %f | CPU: %f |\n", &r.Time, r.SensorId, r.Temperature, r.CPU)
       }
    ```

1.  Formulate an SQL insert statement for the `sensor_data` hypertable:

    ```go
    //SQL query to generate sample data
       queryInsertTimeseriesData := `
       INSERT INTO sensor_data (time, sensor_id, temperature, cpu) VALUES ($1, $2, $3, $4);
       `
    ```

1.  Execute the SQL statement for each sample in the results slice:

    ```go
    //Insert contents of results slice into TimescaleDB
       for i := range results {
           var r result
           r = results[i]
           _, err := dbpool.Exec(ctx, queryInsertTimeseriesData, r.Time, r.SensorId, r.Temperature, r.CPU)
           if err != nil {
               fmt.Fprintf(os.Stderr, "Unable to insert sample into Timescale %v\n", err)
               os.Exit(1)
           }
           defer rows.Close()
       }
       fmt.Println("Successfully inserted samples into sensor_data hypertable")
    ```

> 1.  <Optional />This example `main.go` generates sample data and inserts it into

Reviewer 2: ```suggestion
1.  [](#)<Optional />This example `main.go` generates sample data and inserts it into
```

We'll need the hack whenever we need Markdown to be processed in the same line

       sensorTypes := []string{"a", "a", "b", "b"}
       sensorLocations := []string{"floor", "ceiling", "floor", "ceiling"}

       for i := range sensorTypes {
           //INSERT statement in SQL
           queryInsertMetadata := `INSERT INTO sensors (type, location) VALUES ($1, $2);`

           //Execute INSERT command
           _, err := dbpool.Exec(ctx, queryInsertMetadata, sensorTypes[i], sensorLocations[i])
           if err != nil {
               fmt.Fprintf(os.Stderr, "Unable to insert data into database: %v\n", err)
               os.Exit(1)
           }
           fmt.Printf("Inserted sensor (%s, %s) into database \n", sensorTypes[i], sensorLocations[i])

       }
       fmt.Println("Successfully inserted all sensors into database")

    }
    ```

</Collapsible>

</procedure>

Instead of inserting a single row of data at a time, you can use this procedure
to insert multiple rows of data, instead:

<procedure>

<Collapsible heading="Inserting multiple rows of data" headingLevel={3} defaultExpanded={false}>

1.  This example uses PostgreSQL to generate some sample time-series to insert
    into the `sensor_data` hypertable. Define the SQL statement to generate the

>     data, called `queryDataGeneration`. Then use use the `.Query()` function to

Author: ```suggestion
    data, called `queryDataGeneration`. Then use the `.Query()` function to
```

           os.Exit(1)
       }
       defer rows.Close()
       fmt.Println("Successfully generated sensor data")

       //Store data generated in slice results
       type result struct {
           Time        time.Time
           SensorId    int
           Temperature float64
           CPU         float64
       }
       var results []result
       for rows.Next() {
           var r result
           err = rows.Scan(&r.Time, &r.SensorId, &r.Temperature, &r.CPU)
           if err != nil {
               fmt.Fprintf(os.Stderr, "Unable to scan %v\n", err)
               os.Exit(1)
           }
           results = append(results, r)
       }
       // Any errors encountered by rows.Next or rows.Scan are returned here
       if rows.Err() != nil {
           fmt.Fprintf(os.Stderr, "rows Error: %v\n", rows.Err())
           os.Exit(1)
       }

       // Check contents of results slice
       fmt.Println("Contents of RESULTS slice")
       for i := range results {
           var r result
           r = results[i]
           fmt.Printf("Time: %s | ID: %d | Temperature: %f | CPU: %f |\n", &r.Time, r.SensorId, r.Temperature, r.CPU)
       }
    ```

> 1.  Formulate a SQL insert statement for the `sensor_data` hypertable:

Author: ```suggestion
1.  Formulate an SQL insert statement for the `sensor_data` hypertable:
```

       defer rows.Close()
       fmt.Println("Successfully executed query")

       //Do something with the results of query
       // Struct for results
       type result2 struct {
           Bucket time.Time
           Avg    float64
       }

       // Print rows returned and fill up results slice for later use
       var results []result2
       for rows.Next() {
           var r result2
           err = rows.Scan(&r.Bucket, &r.Avg)
           if err != nil {
               fmt.Fprintf(os.Stderr, "Unable to scan %v\n", err)
               os.Exit(1)
           }
           results = append(results, r)
           fmt.Printf("Time bucket: %s | Avg: %f\n", &r.Bucket, r.Avg)
       }
       // Any errors encountered by rows.Next or rows.Scan are returned here
       if rows.Err() != nil {
           fmt.Fprintf(os.Stderr, "rows Error: %v\n", rows.Err())
           os.Exit(1)
       }

    }
    ```

</Collapsible>

</procedure>

## Next steps

Now that you're able to connect, read, and write to a TimescaleDB instance from
your Go application, be sure to check out these advanced TimescaleDB tutorials:

*   Refer to the [pgx documentation][pgx-docs] for more information about pgx.

> *   Get up and running with TimescaleDB with our [Getting Started][getting-started]

Author: ```suggestion
*   Get up and running with TimescaleDB with the [Getting Started][getting-started]
```

[pgx-docs]: https://pkg.go.dev/github.com/jackc/pgx
[pgx-driver-github]: https://github.com/jackc/pgx

> [timescaledb-install]: /install/latest/

Author: ```suggestion
[install]: /install/latest/
```
